### ISSUE TITLE
fn:parse-uri: Observations (cont.)

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#1381]
The rules for `fn:parse-uri` say:

> If the fragment is the empty string, it is discarded and the fragment is the empty sequence.

However, the result of [fn-parse-uri-043](https://github.com/qt4cg/qt4tests/blob/4659dfdcc92ac57c9a7f482a1ec9f3f509020f52/fn/parse-uri.xml#L746-L762) contains an empty fragment string:

```xml
<test-case name="fn-parse-uri-043">
    <description>Parses a URI ending in "#"</description>
    <created by="Michael Kay" on="2023-10-27"/>
    <test>fn:parse-uri("https://qt4cg.org/specifications/xpath-functions-40/Overview.html#")</test>
    <result>
      <assert-deep-eq>map {
        "authority": "qt4cg.org",
        "fragment": "",
        "host": "qt4cg.org",
        "path": "/specifications/xpath-functions-40/Overview.html",
        "path-segments": ("", "specifications", "xpath-functions-40", "Overview.html"),
        "scheme": "https",
        "hierarchical": true(),
        "uri": "https://qt4cg.org/specifications/xpath-functions-40/Overview.html#"
        }</assert-deep-eq>
    </result>
  </test-case>
```

I assume the tests should be updated?

With regard to the spec, I would suggest adding the normalization rule to query parameters:

> If the *query* is the empty string, it is discarded and the *query* is the empty sequence.

**Comment by ndw** [comment]
Yes. The first issue is an implementation bug that I've now fixed. I'll push new tests in a moment.

Also, I updated the PR for #1320 to include this fix for query strings.

**Comment by ChristianGruen** [comment]
@ndw Fine! Three more observations:

1. The spec for [fn:parse-uri](https://qt4cg.org/specifications/xpath-functions-40/Overview.html#func-parse-uri) contains the following example:

```xquery
parse-uri("tel:+1-816-555-1212")
```

The expected result is…

```xquery
map {
  "hierarchical": false(),
  "path": "+1-816-555-1212",
  "path-segments": "1-816-555-1212",
  "scheme": "tel",
  "uri": "tel:+1-816-555-1212"
}
```

The spec mandates that URI decoding is applied on each token of the path segments. This way, `+1-816-555-1212` is converted to ` 1-816-555-1212`. The result in the spec doesn’t contain the leading space, though. How do you handle this in your implementation?

2. An example in the spec mentions `FOXX0000`, but this error code is not defined.

3. Finally, the result of the generated test case `fo-test-fn-parse-uri-004` is flawed, it cannot be parsed. I assume that’s something Michael has to look at?

**Comment by ndw** [comment]
1. That's a typo in the spec; I probably transformed the output to pretty print it or something and accidentally lost the space. You can see the space in the expected result in the test suite.
2. That was intended as a placeholder. I've never worked out how we invent error codes. Suggest a code and I'll use it! :-)
3. I guess so. I assume that's related to the presence of `&amp;` in the result. I made a naive attempt to just wrap text-only deep-equal bodies in backticks, but it cause havoc. @michaelhkay ?

**Comment by ChristianGruen** [comment]
> That was intended as a placeholder. I've never worked out how we invent error codes. Suggest a code and I'll use it! :-)

🤔 You could choose `FOUR0002`.

**Comment by ChristianGruen** [comment]
@ndw Two more observations (maybe the last?):

1. The result structure of the function does not include the new `absolute` map entry (see »The following map is returned«). As it is binary, I wonder if we should always add it (both `true` and `false`)?
2. The result of various example queries (starting with `parse-uri("https://user@example.com/path/to/file")` and ending with `parse-uri("file://c|/path/to/file")`) still lack the `absolute` map entry.

**Comment by ndw** [comment]
I think I fixed both of those issues in an update to #1440 this morning. Thank you!
